cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -363,25 +381,77 @@ String removeAppVersion(String scheduleId) {
   /**
    * Method to add version in StreamSizeSchedule row key in SchedulerStore.
    *
-   * @throws Exception
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws DatasetManagementException
    */
-  public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    if (isUpgradeComplete()) {
+      LOG.info("{} is already upgraded.", NAME);
+      return;
+    }
+
+    final AtomicInteger maxNumberUpdateRows = new AtomicInteger(1000);
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of {}.", NAME);
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
+          .execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() {
+              if (upgradeVersionKeys(table, maxNumberUpdateRows.get())) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, COLUMN, Bytes.toBytes(ProjectInfo.getVersion().toString()));
+              }
+            }
+          });
+      } catch (TransactionFailureException e) {
+        if (e instanceof TransactionConflictException) {
+          LOG.debug("Upgrade step faced Transaction Conflict exception. Retrying operation now.", e);
+          sleepTimeInMins.set(0);
+        } else if (e instanceof TransactionNotInProgressException) {
+          int currMaxRows = maxNumberUpdateRows.get();
+          if (currMaxRows > 500) {
+            maxNumberUpdateRows.decrementAndGet();
+          }
+          sleepTimeInMins.set(0);
```

Comment:
Better to sleep for some time. Otherwise this will keep retrying and prevent other transaction from happening